fission-web-server/library/Fission/Web/Server/Handler/User/Challenge.hs

```diff
@@ -0,0 +1,39 @@
+module Fission.Web.Server.Handler.User.Challenge (handler) where
+
+import           Servant
+import           Servant.Server.Generic
+
+import           Fission.Prelude
+
+import qualified Fission.Web.API.User.Challenge.Types               as Challenge
+
+import           Fission.Web.Server.Authorization.Types
+import qualified Fission.Web.Server.Challenge.Retriever.Class       as Challenge
+import qualified Fission.Web.Server.Challenge.Verifier.Class        as Challenge
+import qualified Fission.Web.Server.Error                           as Web.Err
+import           Fission.Web.Server.Models
+import qualified Fission.Web.Server.RecoveryChallenge.Creator.Class as RecoveryChallenge
+import           Fission.Web.Server.Redirect
+import           Fission.Web.Server.User.Retriever.Class            as User
+
+handler ::
+  ( Challenge.Retriever       m
+  , Challenge.Verifier        m
+  , RecoveryChallenge.Creator m
+  , User.Retriever            m
+  , MonadThrow                m
+  , MonadLogger               m
+  , MonadTime                 m
+  )
+  => Challenge.Routes (AsServerT m)
+handler = Challenge.Routes {..}
+  where
+    recover username = do
+      Entity userId User { } <- Web.Err.ensureMaybe couldntFindUser =<< getByUsername username
+      now       <- currentTime
+      challenge <- RecoveryChallenge.create userId now
```

Comment:
However, here we don't check any other property about the authentication, right?

So if I read this correctly, doesn't this allow any fission user to issue & get a challenge for any other fission user?

E.g. a malicious actor could create a new fission account, and then use their UCAN to run a request `/user/challenge/recover/bob`, and then use the response to call `/user/did/bob/{challenge}` with a DID they control as the body for that request, essentially taking away my control over bob's fission account and giving it to themselves?

Reply:
Ahh thanks for catching that! I'll admit I have a fairly tenuous understanding of this codebase. This new endpoint was cobbled together from bits of various other ones I saw in the repo 😅 Looking at some other examples, I think I see how to add that validation here. I'll get this fixed today!

Reply:
What's the goal, actually? How should the auth gating work for that endpoint?
Is the plan to require the user to be authed to access this endpoint? Although in that case I wonder why they'd need to recover in the first place.
Is the idea to allow this endpoint only for the email service? I.e. the email service has a special DID that this fission server respects and it opens up this endpoint for it?
Or is the idea that the fission service's DID would be the "root owner" in this case, and we'd check that the root DID of a UCAN coming in to use the recovery endpoint is the fission service's DID, allowing us to delegate a UCAN for accessing the recovery endpoint to other services such as the email service?

Reply:
The current approach we're taking is:
- On registration, the client builds a UCAN that delegates root access to the `rs-email-service`'s DID
- That UCAN is sent to the `rs-email-service` and stored in a DB that maps usernames to UCANS
- Then, when a user goes to recover their account, they enter their username and a request is sent from the client to the `rs-email-service`, which finds the associated UCAN in the DB and uses that to make a request to the this new fission server endpoint to get a challenge
- With the challenge in hand, the `rs-email-service` should be able to call `/v2/api/user/did/{Username}/{Challenge}` to recover the user's account

Reply:
I should add that the `rs-email-service` is just meant to be an example of how this email recovery flow could work. We expect devs to create their own services that maybe reference our `rs-email-service` example

Reply:
That sound awesome and exactly like what I'd expect.

Let's go 🚀